File: src/sdk/schema_system.cpp
#include "sdk/sdk.hpp"

namespace sdk {
    SchemaSystem* SchemaSystem::get() noexcept {
        const auto address = process::find_pattern("schemasystem.dll", "48 8D 0D ? ? ? ? E9 ? ? ? ? CC CC CC CC 48 8D 0D ? ? ? ? E9 ? ? ? ? CC CC CC CC 48 83 EC 28");

        if (!address.has_value())
            return nullptr;

        return address->rip().as<SchemaSystem*>();
    }

    std::vector<SchemaSystemTypeScope*> SchemaSystem::type_scopes() const noexcept {
        const auto count = process::read_memory<std::uint32_t>(reinterpret_cast<std::uint64_t>(this) + 0x190);
        const auto data = process::read_memory<std::uint64_t>(reinterpret_cast<std::uint64_t>(this) + 0x198);

        if (count == 0 || data == 0)
            return {};

        std::vector<sdk::SchemaSystemTypeScope*> type_scopes;

        type_scopes.resize(count);

        process::read_memory(data, type_scopes.data(), count * sizeof(std::uint64_t));

        return type_scopes;
    }
}

File: src/sdk/schema_system.rs
use std::mem;

use anyhow::{bail, Result};

use super::SchemaSystemTypeScope;

use crate::os::Process;

use crate::config::SCHEMA_CONF;

pub struct SchemaSystem<'a> {
    process: &'a Process,
    address: usize,
}

impl<'a> SchemaSystem<'a> {
    pub fn new(process: &'a Process) -> Result<Self> {
        let mut address = process
            .find_pattern(SCHEMA_CONF.module_name, SCHEMA_CONF.pattern)
            .expect("unable to find schema system pattern");

        address = process.resolve_rip(address, None, None)?;

        Ok(Self { process, address })
    }

    pub fn type_scopes(&self) -> Result<Vec<SchemaSystemTypeScope>> {
        let size = self
            .process
            .read_memory::<u32>(self.address + SCHEMA_CONF.type_scope_size_offset)?;

        if size == 0 {
            bail!("no type scopes found");
        }

        let data = self
            .process
            .read_memory::<usize>(self.address + SCHEMA_CONF.type_scope_data_offset)?;

        let mut addresses = vec![0; size as usize];

        self.process.read_memory_raw(
            data.into(),
            addresses.as_mut_ptr() as *mut _,
            addresses.len() * mem::size_of::<usize>(),
        )?;

        let type_scopes: Vec<SchemaSystemTypeScope> = addresses
            .iter()
            .map(|&address| SchemaSystemTypeScope::new(self.process, address))
            .collect();

        Ok(type_scopes)
    }
}
